### PR TITLE
Feat/long-term schduler

### PR DESCRIPTION
**Pull request by ishiko732**
@L-M-Sherlock 
I am uncertain whether the state transition from New to Review is correct; I need you to verify the `LongTermScheduler.newState` and `LongTermScheduler.reviewState` methods.

### CODE REVIEW

**Comment by L-M-Sherlock** [comment]
How can I check them quickly? Do you have any unit tests?

File: src/fsrs/impl/long_term_schduler.ts [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    easy_interval = this.algorithm.next_interval(next_easy.stability, interval)
+
+    again_interval = Math.min(again_interval, hard_interval) as int
+    hard_interval = Math.max(hard_interval, again_interval + 1) as int
+    good_interval = Math.max(good_interval, hard_interval + 1) as int
+    easy_interval = Math.max(easy_interval, good_interval + 1) as int
+
+    next_again.scheduled_days = again_interval
+    next_again.due =
+      again_interval > 0
```
Comment (L-M-Sherlock) [inline src/fsrs/impl/long_term_schduler.ts]: When would it be less or equal to zero?

File: src/fsrs/default.ts [hunk 2]
```diff
@@ -29,7 +30,9 @@ export const generatorParameters = (
     maximum_interval: props?.maximum_interval || default_maximum_interval,
     w: w,
     enable_fuzz: props?.enable_fuzz || default_enable_fuzz,
-  }
+    enable_short_term:
+      props?.enable_short_term || defualt_enable_short_term,
```
Comment (ishiko732) [inline src/fsrs/default.ts]:
```typescript
    enable_short_term:
      props?.enable_short_term ?? defualt_enable_short_term,
```

File: src/fsrs/impl/long_term_schduler.ts (hunk 1 above)
Comment (ishiko732) [reply src/fsrs/impl/long_term_schduler.ts]: > When would it be less or equal to zero?

I have reviewed the code and confirmed that no cases of <=0 exist; this was my mistake.

https://github.com/open-spaced-repetition/ts-fsrs/blob/a752a6e172fb18879e710d008d5fbb4949bb8730/src/fsrs/algorithm.ts#L154-L164

**Comment by ishiko732** [comment]
> Do you have any unit tests?

https://github.com/open-spaced-repetition/ts-fsrs/blob/11429fff6568dc49ed2d549f2b751bfdecc7c5c9/__tests__/impl/long-term_schduler.test.ts

```typescript
  const parameters = {
    request_retention: 0.9,
    maximum_interval: 36500,
    w: [
      0.4197, 1.1869, 3.0412, 15.2441, 7.1434, 0.6477, 1.0007, 0.0674, 1.6597,
      0.1712, 1.1178, 2.0225, 0.0904, 0.3025, 2.1214, 0.2498, 2.9466, 0.4891,
      0.6468,
    ],
    enable_fuzz: false,
    enable_short_term: false,
  }
```
first_time: 2022/11/29 12:30:00

## test1
Grades: 
```typescript
[Good, Good, Good, Good, Good, Good, Again, Again, Good, Good, Good, Good, Good]
```
ivl_history: 
```
[3, 6, 17, 42, 95, 200, 8, 2, 3, 3, 6, 10, 17]
```
s_history: 
```typescript
[
      0.57587467, 6.28341418, 16.83356103, 41.95128557, 95.07063986,
      199.53765138, 8.31519008, 1.59859456, 1.59859456, 3.49350932, 5.58821385,
      9.83641092, 16.95422162,
    ]
```
```typescript
d_history:
[
      7.1434, 7.1434, 7.1434, 7.1434, 7.1434, 7.1434, 9.00990564, 10,
      9.80746516, 9.62790717, 9.46045139, 9.30428213, 9.15863867,
    ]
```

## test2
Grades: 
```typescript
[Again, Hard, Good, Easy, Again, Hard, Good, Easy]
```
ivl_history: 
```
[1, 2, 3, 8, 2, 2, 3, 8]
```
s_history: 
```typescript
[
      0.21652154, 0.21652154, 1.32177381, 7.69767996, 1.91650418, 1.91650418,
      3.02442382, 8.45226282,
    ]
```
```typescript
d_history:
[
      9.00990564, 9.81735598, 9.63713135, 8.53580104, 10, 10, 9.80746516,
      8.69465435,
    ]
```

## test3
Grades: 
```typescript
[Hard, Good, Easy, Again, Hard, Good, Easy, Again]
```
ivl_history: 
```
[2, 3, 17, 3, 3, 5, 15, 3]
```
s_history: 
```typescript
[
      0.35311368, 3.40179546, 16.86596974, 2.9223039, 2.9223039, 4.91748183,
      15.09437624, 2.71341838,
    ]
```
```typescript
d_history:
[
      8.07665282, 8.01375158, 7.02183706, 8.89653604, 9.71162749, 9.53852896,
      8.44384445, 10,
    ]
```

## test4
Grades: 
```typescript
[Good, Easy, Again, Hard, Good, Easy, Again, Hard]
```
ivl_history: 
```
[3, 17, 3, 3, 6, 23, 3, 3]
```
s_history: 
```typescript
[
      0.57587467, 17.3937106, 2.98322161, 2.98322161, 6.06697209, 23.2008818,
      3.36166219, 3.36166219,
    ]
```
```typescript
d_history:
[
      7.1434, 6.21014718, 8.13955406, 9.0056661, 8.88014936, 7.82983963,
      9.65007924, 10,
    ]
```

## test5
Grades: 
```typescript
[Easy, Again, Hard, Good, Easy, Again, Hard, Good]
```
ivl_history: 
```
[4, 1, 2, 3, 13, 3, 3, 4]
```
s_history: 
```typescript
[
      0.93916394, 0.70772253, 0.70772253, 2.9537906, 12.61933619, 2.51177108,
      2.51177108, 4.07590188,
    ]
```
```typescript
d_history:
[
      6.21014718, 8.13955406, 9.0056661, 8.88014936, 7.82983963, 9.65007924, 10,
      9.80746516,
    ]
```

**Comment by ishiko732** [comment]
> How can I check them quickly? 

I will write a preview page when I have time.

**Comment by L-M-Sherlock** [comment]
> ## test1
> Grades:
> 
> ```ts
> [Good, Good, Good, Good, Good, Good, Again, Again, Good, Good, Good, Good, Good]
> ```
> 
> ivl_history:
> 
> ```
> [3, 6, 17, 42, 95, 200, 8, 2, 3, 3, 6, 10, 17]
> ```
> 
> s_history:
> 
> ```ts
> [
>       0.57587467, 6.28341418, 16.83356103, 41.95128557, 95.07063986,
>       199.53765138, 8.31519008, 1.59859456, 1.59859456, 3.49350932, 5.58821385,
>       9.83641092, 16.95422162,
>     ]
> ```
> 
> ```ts
> d_history:
> [
>       7.1434, 7.1434, 7.1434, 7.1434, 7.1434, 7.1434, 9.00990564, 10,
>       9.80746516, 9.62790717, 9.46045139, 9.30428213, 9.15863867,
>     ]
> ```

Why the first `Good` after the last `Again` doesn't change the stability?

**Comment by ishiko732** [comment]
> Why the first `Good` after the last `Again` doesn't change the stability?

I don’t understand why all the results from the `next_recall_stability` calculation are exactly the same.
image

File: src/fsrs/impl/long_term_schduler.ts [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    good_interval = this.algorithm.next_interval(next_good.stability, interval)
+    easy_interval = this.algorithm.next_interval(next_easy.stability, interval)
+
+    again_interval = Math.min(again_interval, hard_interval) as int
+    hard_interval = Math.max(hard_interval, again_interval + 1) as int
+    good_interval = Math.max(good_interval, hard_interval + 1) as int
+    easy_interval = Math.max(easy_interval, good_interval + 1) as int
+
+    next_again.scheduled_days = again_interval
+    next_again.due = this.review_time.scheduler(again_interval as int)
```
Comment (ishiko732) [inline src/fsrs/impl/long_term_schduler.ts]:
```typescript
    next_again.scheduled_days = again_interval
    next_again.due = this.review_time.scheduler(again_interval as int, true)
```

**Comment by ishiko732** [comment]
> Why the first `Good` after the last `Again` doesn't change the stability?

https://github.com/open-spaced-repetition/ts-fsrs/blob/f838a179f9371e8acf8703a028aa04995f5cdfa3/src/fsrs/impl/long_term_schduler.ts#L213-L214
scheduled it by minutes here, which was a mistake. I’ll make the adjustments tonight. 😢

**Comment by ishiko732** [comment]
## switch long-term schduler test

Grades: 
```typescript
[Good(short), Good(long), Again(long), Good(long), Good(short), Again(short)]
```
ivl_history: 
```
[0, 4, 1, 4, 12, 0]
```

s_history: 
```typescript
[
      3.0412, 3.0412, 1.20788692, 3.83856852, 12.23542321, 2.48288917,
    ]
```

d_history:
```typescript
[4.49094334, 4.66971892, 6.70295066, 6.73263695,6.76032238,8.65264745]
```

state_history:
```typescript
['Learning', 'Review', 'Review', 'Review','Review','Relearning']
```

**Comment by L-M-Sherlock** [comment]
Hi, @remnoteio, @sobjornstad, @bjsi. This PR will allow user to skip the short-term schedule. I guess you will be interested in it.

**Review by L-M-Sherlock — APPROVED** [review]
LGTM

**Review by L-M-Sherlock — APPROVED** [review]
LGTM

File: src/fsrs/impl/long_term_schduler.ts [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+
+export default class LongTermScheduler extends AbstractScheduler {
+  protected override newState(grade: Grade): RecordLogItem {
+    const exist = this.next.get(grade)
+    if (exist) {
+      return exist
+    }
+
+    const first_difficulty = this.algorithm.init_difficulty(grade)
+    const first_stability = this.algorithm.init_stability(grade)
```
Comment (ishiko732) [inline src/fsrs/impl/long_term_schduler.ts:20]: Each `grade` should have its corresponding `stability` and `difficulty` set.

Comment (L-M-Sherlock) [reply src/fsrs/impl/long_term_schduler.ts:20]: Before the first learning, the card doesn't have stability and difficulty. In the first learning, each grade has its corresponding w[0:4]. What's the problem here?

Comment (ishiko732) [reply src/fsrs/impl/long_term_schduler.ts:20]: > Before the first learning, the card doesn't have stability and difficulty. In the first learning, each grade has its corresponding w[0:4]. What's the problem here?

The `preview` in `abstract_scheduler.ts` collects results for all grades [Again, Good, Hard, Easy], and after the first call to `review`, it stores the results, causing all to have `init_stability` and `init_difficulty` set according to `Rating.Again`.
https://github.com/open-spaced-repetition/ts-fsrs/blob/d43b6b006e24875550ca3ce842958598fb2c8679/src/fsrs/abstract_schduler.ts#L48-L74

Comment (L-M-Sherlock) [reply src/fsrs/impl/long_term_schduler.ts:20]: Where does the `again` come from?

**Pull request merged**